src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,276 @@
+/// </code>
+/// The validation bytes are followed by the checksum.  The checksum is needed in the message as assets can be found
+/// in any order (they are not reported in the order of the array of checksums passed into the writing method).
+/// Following this is the kind of the asset.  This kind is used by the reading code to know which
+/// asset-deserialization routine to invoke. Finally, the asset data itself is written out.
+/// </summary>
+internal static class RemoteHostAssetSerialization
 {
-    internal static class RemoteHostAssetSerialization
+    /// <summary>
+    /// A sentinel byte we place between messages.  Ensures we can detect when something has gone wrong as soon as possible.
+    /// </summary>
+    private const byte MessageSentinelByte = 0b01010101;
+
+    private static readonly ObjectPool<SerializableBytes.ReadWriteStream> s_streamPool = new(SerializableBytes.CreateWritableStream);
+
+    public static async ValueTask WriteDataAsync(
+        PipeWriter pipeWriter,
+        AssetPath assetPath,
+        ReadOnlyMemory<Checksum> checksums,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
     {
-        public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
-            ReadOnlyMemory<Checksum> checksums,
-            CancellationToken cancellationToken)
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(new UnboundedChannelOptions()
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            // We have a single task reading the data from the channel and writing it to the pipe.  This option
+            // allows the channel to operate in a more efficient manner knowing it won't have to sychronize data
+            // for multiple readers.
+            SingleReader = true,
+
+            // Currently we only have a single writer writing to the channel when we call FindAllAssetsAsync.
+            // However, we could change this in the future to allow the search to happen in parallel.
+            SingleWriter = true,
+        });
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe.
+        using var _ = cancellationToken.Register(() => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
```

Comment:
>() =>  [](http://example.com/codeflow?start=49&length=6)

Looks like there is a version of Register that takes in argument to pass to the callback. Unfortunately, that argument is an object, so I'm not sure whether getting an object from a pool and populating it with the channel is better than the closure allocation.

Reply:
Yup.  You allocate no matter what.

Reply:
I guess the question I had is whether it would be worthwhile to pool an object that holds the channel to avoid that allocation.

Reply:
Or, rather than even needing a pool, couldn't the channel just be sent as the object param as the CancellationToken is already sent?

Reply:
Or is that version of the Register method not available in the runtime where this is executed?

Reply:
done.